macros/src/trans.rs

```diff
@@ -157,14 +157,23 @@ fn init(app: &App, main: &mut Vec<Tokens>, root: &mut Vec<Tokens>) {
 
     let mut tys = vec![quote!(#device::Peripherals)];
     let mut exprs = vec![quote!(#device::Peripherals::all())];
+    let mut ret = None;
     let mut mod_items = vec![];
 
-    if !app.resources.is_empty() {
+    // Write resources usable by `init`, if any
+
+
+    // Are there any resources that have an initializer? Those can be used by `init`.
+    let has_initialized_resources = app.resources.iter()
+        .find(|&(_, res)| res.expr.is_some()).is_some();
+
+    if has_initialized_resources {
         let mut fields = vec![];
         let mut lifetime = None;
         let mut rexprs = vec![];
 
-        for (name, resource) in &app.resources {
+        for (name, resource) in app.resources.iter()
+                .filter(|&(_, res)| res.expr.is_some()) {
```

Comment:
hmm, I wonder if this would be more "correct" as `app.resources.iter().filter_map(|(name, res)| res.map(|res| (name, rers)))`. It does look more complicated though. Basically what I'm trying to do is avoid having `resource` be an "Option" that it's always Some. In any case, we won't unwrap the Option in this branch so maybe I'm just being pendantic ...